ethcore/src/client/client.rs

```diff
@@ -122,11 +122,13 @@ impl SleepState {
 /// Call `import_block()` to import a block asynchronously; `flush_queue()` flushes the queue.
 pub struct Client {
 	mode: Mode,
-	chain: Arc<BlockChain>,
-	tracedb: Arc<TraceDB<BlockChain>>,
+	chain: RwLock<Arc<BlockChain>>,
```

Comment:
I would favor giving `Client` full ownership of `BlockChain`, only loaning it out temporarily to the `TraceDB`.